tests/unit/test_endpoint_deduplication.py

```diff
@@ -0,0 +1,57 @@
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import json
+
+import yaml
+from ops.testing import Container, Context, Relation, State
+
+from charm import GrafanaAgentK8sCharm
+from grafana_agent import CONFIG_PATH
+
+
+def test_prometheus_endpoints_deduplication():
+    # GIVEN a set of duplicate Prometheus endpoints
+    remote_write_url = "http://traefik.ip/cos-mimir/api/v1/push"
+    mimir_relation = Relation(
+        "send-remote-write",
+        remote_units_data={
+            0: {"remote_write": json.dumps({"url": remote_write_url})},
+            1: {"remote_write": json.dumps({"url": remote_write_url})},
+            2: {"remote_write": json.dumps({"url": remote_write_url})},
+        },
+    )
+    state = State(
+        leader=True,
+        containers={Container(name="agent", can_connect=True)},
+        relations=[
+            mimir_relation,
+        ],
+        config={
+            "forward_alert_rules": True,
+        },
+    )
+
+    # WHEN the charm processes endpoints with duplicates
+    ctx = Context(charm_type=GrafanaAgentK8sCharm)
+
+    with ctx(ctx.on.config_changed(), state) as mgr:
```

Comment:
you don't need to use the context manager for this test. 
```suggestion
    output_state = ctx.run(ctx.on.config_changed(), state)
    <unindent the rest>
```

Reply:
Great, done!